### ISSUE TITLE
Correct format for passing files as command line variables for fuzzing

### ISSUE THREAD

**Issue by Mikerah** [AFLplusplus/AFLplusplus#1358]
I'm trying to fuzz a program that takes as inputs a series of files.
I have been able to follow the instructions outlined [here](https://github.com/AFLplusplus/AFLplusplus/tree/stable/utils/argv_fuzzing). However, I noticed that once I take a look at the input test cases generated by AFL++ and run a back trace in gdb, the problem is not reading right file.
Here's an example input:
```
Program received signal SIGSEGV, Segmentation fault.
__memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:390
Download failed: Function not implemented.  Continuing without source file ./string/../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S.
390	../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
(gdb) bt
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:390
#1  0x00005555555aaff4 in memcpy (__len=4096, __src=0xffffffffffffffff, __dest=<optimized out>)
    at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29
#2  BinFileUtils::BinFile::BinFile (this=0x55555566c670, fileName=..., _type="zkey", maxVersion=1)
    at ../src/binfile_utils.cpp:30
#3  0x00005555555ac2ac in BinFileUtils::openExisting (filename="/", type="zkey", 
    maxVersion=maxVersion@entry=1) at ../src/binfile_utils.cpp:143
#4  0x0000555555560272 in main (argc=<optimized out>, argv=0x5555555dc3c0 <afl_init_argv(int*)::ret>)
    at ../src/main_prover.cpp:42

```
Noticed that at line 2 and 3, the program is reading `/` and `...` when as inputs, I'm passing in the following
```
./build/prover\x00../../circom-test-cases/snarkjs-test-case/circuit_groth16_0.zkey\x00../../circom-test-cases/snarkjs-test-case/witness.wtns\x00../../circom-test-cases/snarkjs-test-case/proof.json\x00../../circom-test-cases/snarkjs-test-case/public.json\x00
```
So, `argv[1]` should be `../../circom-test-cases/snarkjs-test/case/circuit_groth16_0.zkey` and so on.

What is the appropriate way to construct my test case file so that it can take as input a file name?

**Comment by vanhauser-thc** [comment]
you do not show the afl command line, not what you set in the env var ... and in general there is not enough context for me to even guess what your issue is. please put in more effort if you want help.

**Comment by Mikerah** [comment]
I simply ran the basic `afl-fuzz` as per the instructions. In particular I ran `afl-fuzz -i input -o output ./binary` where my input file has a single test case which is ```./build/prover\x00../../circom-test-cases/snarkjs-test-case/circuit_groth16_0.zkey\x00../../circom-test-cases/snarkjs-test-case/witness.wtns\x00../../circom-test-cases/snarkjs-test-case/proof.json\x00../../circom-test-cases/snarkjs-test-case/public.json\x00```.

After letting the fuzzer run for a while, until I got some crashes in `out/defaults/crashes`. When I run one of the crashing inputs (e.g. `id:000000`) in a gdb instance, I get a segfault as expected. Upon further inspection as noted in the OP, it turns out it's because the wrong input is getting passed into the binary.

For some more context, the binary takes as input 4 files on the command line. These files reside in another directory.
I'm asking if this is indeed the correct way to go about trying to fuzz a command line program that takes as inputs files.

**Comment by vanhauser-thc** [comment]
afl-fuzz can only mutate one input for fuzzing, not multiple files.
you can specify with -f /path/to/file  if the file to fuzz has a specific location.
still not grasping the whole context of what you want to do ...

**Comment by Mikerah** [comment]
I want to fuzz a binary that takes as input 4 files as part of its normal operation. I modified the code in which I added `AFL_INIT_ARGV()` at the beginning of the main function (along with the required import) in order to be able to enable AFL++ to read from the argv array. I am able to get the fuzzer to run when I make these changes. However, once I look at the actual inputs generated by AFL++, I noticed that my files are not being used as inputs as per the OP.

Do you know if there is a way to construct the input file to AFL++ so that it can take as input 4 files for such a program?

**Comment by Mikerah** [comment]
@vanhauser-thc Let me rephrase my problem:

I have a program (whose source code I have access to, if that matters) that takes as input 4 files at the command line. None of these files are optional parameters, they are required. As stated before, I followed the instructions to enable argv fuzzing and created a testcase where I pass in 4 example files for the program. 

After running the fuzzer for a while, I got a few results. All of the results were segmentation faults and upon further examination, it was at the file reading step for the program. As such, the core meat of the program didn't really get touched at all during my fuzzing attempts. It's clearly because by doing argv fuzzing, the fuzzer is only mutating the argv inputs and since the program is expecting to be able to read these inputs as files, it crashes before it gets anywhere into the program.

My current options are modify the source directly so that it takes in the contents of each file directly and forgo argv fuzzing OR somehow get argv fuzzing to do what I expect it to do.

Please advise.

**Comment by vanhauser-thc** [comment]
it is always best to modify the source code to have an effective way of fuzzing a target. argv fuzzing is for many targets pointless. if the target has several features you can create different fuzzer setups for them.

**Comment by Mikerah** [comment]
Thank you for the advice @vanhauser-thc. I will try to modify the source code directly.
